### PR TITLE
feat: do not close commit message with `q` until saved

### LINKED ISSUE #552: `q` should not abort the commit if it contains a written commit message.
I've encountered this multiple times because I try `:wq` to save and quit to commit, but accidentally just hit `q` which makes me lose the commit message.

Having a prompt if there are unsaved changes would be preferred

### PR DESCRIPTION
**Pull request by treatybreaker**
Prior to this commit hitting `q` in normal mode in the commit editor would close it even if the text had not been saved. Also, prior to this commit, the `Buffer:get_option` in `lib/buffer.lua` never actually returned the option value :sweat_smile:. 

This PR pops up a warning when `q` is hit in normal mode and the buffer has unsaved modifications, we may want to change the message and the log level. 
Currently the message is: 
> Commit message has not been saved! Try `:w`.

 and the log level is set to `WARN`.

This PR also fixes the `Buffer:get_option` to *actually* return the value it got.

Let me know if any changes are desired to the message or log level, and secondly I have a question. Do y'all prefer me to force push or to create a separate commit for each change? Usually I try to avoid polluting the git history with a bunch of teeny tiny fix commits, but y'all may prefer that.

Closes #552

### CODE REVIEW

**Comment by CKolkey** [comment]
First off, hats off to you for all the recent work! And for fixing that buffer method. Oops. 

Alright, lets get into this. I think you can combine your efforts with what I started here: https://github.com/NeogitOrg/neogit/pull/567

Essentially, instead of nudging a user to save, just ask if they want to abort the commit, because that should be allowed. But I like your way to check if the buffer is modified more, so definitely keep that.

And don't worry about the `client.wrap()` stuff - thats out of scope for this, and part of a bigger cleanup, so I'll do that in another branch.


As for the commits, thats up to you. A reasonably clean history is nice, without out too many "lint" commits, but... I'm also guilty of those sometimes, so don't worry about it too much.

**Comment by treatybreaker** [comment]
1. Thanks, I've got a bunch of time on my hands and I got tired of writing Rust in the last week.

2. I didn't notice #567, are you sure you want confirmation dialogues for it?
Would it be better to just inform the user to hit `:q!`? Dialogues block the editor and don't conform to standard keymaps. If the confirmation is preferred then I can certainly integrate it over here.

3. Then I'll do my standard commit stuff of force pushing for tiny chunks and new commits for reasonably new stuff/significant modifications.

**Comment by treatybreaker** [comment]
Oh wait, yeah a dialogue is definitely better. They're hitting `q` in normal mode, not doing anything funky. Will integrate.

**Comment by treatybreaker** [comment]
@CKolkey 

Alright, take a look again. I removed the `write` variable and check the modified status as well as getting user input.

As a heads up `nvim_buf_get_option` is deprecated in `0.10`, so that's something we have to keep an eye out for. I don't want to write the shims for compatibility in this PR, but it will need to be done prior to `0.11` or `0.12`, whenever they deprecate it.

**Review by ten3roberts — APPROVED** [review]
I like the state of this so far, and the solution is really nice and goes out of your way without trying to do too much surprising side effects behind the scenes (such as saving somewhere)

File: lua/neogit/buffers/commit_editor/init.lua [hunk 1]
```diff
@@ -30,7 +30,7 @@ function M.new(filename, on_unload)
 end
 
 function M:open()
-  local written = false
+  local commit_aborted = false
```
[the remark this answers is not in the document]
Comment (treatybreaker) [reply lua/neogit/buffers/commit_editor/init.lua:33]: You're right, I usually try to do positive action names. Slipped my mind here. Will fix.

[the remark this answers is not in the document]
Comment (CKolkey) [reply lua/neogit/buffers/commit_editor/init.lua:33]: I agree, it's a little convoluted to be a double negative

[the remark this answers is not in the document]
Comment (treatybreaker) [reply lua/neogit/buffers/commit_editor/init.lua]: Fixed in the latest force push, using the name `should_commit`. Take a look and let me know if that looks good @ten3roberts.

File: lua/neogit/buffers/commit_editor/init.lua [hunk 2]
```diff
@@ -40,25 +40,25 @@ function M:open()
...
     readonly = false,
     autocmds = {
-      ["BufWritePre"] = function()
-        written = true
-      end,
       ["BufUnload"] = function(o)
-        if written then
+        local buf = Buffer.create {
+          name = o.buf,
+        }
```
Comment (CKolkey) [inline lua/neogit/buffers/commit_editor/init.lua:46]: Could you explain what this is doing? Not sure I follow the logic. It's creating a new buffer to check the modified status?

Comment (treatybreaker) [reply lua/neogit/buffers/commit_editor/init.lua:46]: So previously we were directly interfacing with the nvim api to manage this `o.buf`, but we already have a built-in buffer class that has a bunch of these management functions.

I'm constructing our internal Buffer class so we can use in-built functionality from our own lib.

**Pull request merged**